gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/MultiTaskSession.java

```diff
@@ -0,0 +1,283 @@
+     */
+    public MultiTaskSession(final SessionTask initialSessionTask, final String sessionId,
+                     final ConcurrentMap<String, Session> sessions) {
+        super(initialSessionTask, sessionId, false, sessions);
+
+        queue = new LinkedBlockingQueue<>(initialSessionTask.getSettings().maxSessionTaskQueueSize);
+
+        // using a global function cache is cheaper than creating a new on per session especially if you have to
+        // create a lot of sessions. it will generate a ton of throw-away objects. mostly keeping the option open
+        // to not use it to preserve the ability to use the old functionality if wanted or if there is some specific
+        // use case with sessions that needs it. if we wanted this could eventually become a per-request option
+        // so that the client could control it as necessary and get scriptengine isolation if they need it.
+        if (initialSessionTask.getSettings().useCommonEngineForSessions)
+            scriptEngineManager = initialSessionTask.getGremlinExecutor().getScriptEngineManager();
+        else
+            scriptEngineManager = initializeGremlinExecutor(initialSessionTask).getScriptEngineManager();
+
+        scheduledExecutorService = initialSessionTask.getScheduledExecutorService();
+        submitTask(initialSessionTask);
+    }
+
+    /**
+     * Gets the script engine specific to this session which is dependent on the
+     * {@link Settings#useCommonEngineForSessions} configuration.
+     */
+    @Override
+    public GremlinScriptEngine getScriptEngine(final SessionTask sessionTask, final String language) {
+        return scriptEngineManager.getEngineByName(language);
+    }
+
+    @Override
+    public boolean isAcceptingTasks() {
+        return !ending.get();
+    }
+
+    @Override
+    public boolean submitTask(final SessionTask sessionTask) throws RejectedExecutionException {
+        try {
+            return isAcceptingTasks() && queue.add(sessionTask);
+        } catch (IllegalStateException ise) {
+            final String msg = String.format("Task %s rejected from session %s",
+                    sessionTask.getRequestMessage().getRequestId(), getSessionId());
+            throw new RejectedExecutionException(msg);
+        }
+    }
+
+    @Override
+    public void run() {
+        // allow the Session to know about the thread that is running it - the thread really only has relevance
+        // once the session has started.
+        this.sessionThread = Thread.currentThread();
+
+        // there must be one item in the queue at least since addTask() gets called before the worker
+        // is ever started
+        SessionTask sessionTask = queue.poll();
+        if (null == sessionTask)
+            throw new IllegalStateException(String.format("Worker has no initial context for session: %s", getSessionId()));
+
+        try {
+            startTransaction(sessionTask);
```

Comment:
I don't completely understand this. Wouldn't this start a new trx for every request in the same session? Is that expected behaviour?

Reply:
the answer to this might be more clear given my improved comments here: https://github.com/apache/tinkerpop/pull/1414/files#r611785656 

but basically, that method is only called when the session is started.